File: FILE.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <unordered_map>
#include <algorithm>
#include <cmath>

// Define data structures
struct Email {
    std::string MetadataFrom;
    std::string MetadataTo;
    std::string RawText;
};

struct Alias {
    std::string Alias;
    int PersonId;
};

struct Person {
    int Id;
    std::string Name;
};

struct Edge {
    std::string sender;
    std::string receiver;
    int weight;
};

// Function to unify names and aliases
std::string unifyName(const std::string& name, const std::unordered_map<std::string, int>& aliasMapping, const std::unordered_map<int, std::string>& personMapping) {
    std::string unifiedName = name;
    std::transform(unifiedName.begin(), unifiedName.end(), unifiedName.begin(), ::tolower);
    size_t atPos = unifiedName.find('@');
    if (atPos != std::string::npos) {
        unifiedName = unifiedName.substr(0, atPos);
    }
    auto aliasIt = aliasMapping.find(unifiedName);
    if (aliasIt != aliasMapping.end()) {
        unifiedName = personMapping.at(aliasIt->second);
    }
    return unifiedName;
}

// Function to calculate PageRank
std::map<std::string, double> calculatePageRank(const std::vector<Edge>& edges) {
    std::map<std::string, double> pageRank;
    std::map<std::string, double> newPageRank;
    const double dampingFactor = 0.85;
    const double initialRank = 1.0 / edges.size();

    // Initialize PageRank values
    for (const Edge& edge : edges) {
        pageRank[edge.sender] = initialRank;
        pageRank[edge.receiver] = initialRank;
    }

    // Perform PageRank iterations
    for (int iteration = 0; iteration < 10; ++iteration) {
        for (const Edge& edge : edges) {
            newPageRank[edge.receiver] += dampingFactor * pageRank[edge.sender] / edges.size();
        }
        pageRank = newPageRank;
        newPageRank.clear();
    }
    return pageRank;
}

int main() {
    // Load the emails dataset
    std::ifstream emailsFile("Emails.csv");
    std::vector<Email> emails;
    // Read emails data into 'emails' vector

    // Load the aliases dataset
    std::ifstream aliasesFile("Aliases.csv");
    std::vector<Alias> aliases;
    // Read aliases data into 'aliases' vector

    // Load the persons dataset
    std::ifstream personsFile("Persons.csv");
    std::vector<Person> persons;
    // Read persons data into 'persons' vector

    // Create alias and person mappings
    std::unordered_map<std::string, int> aliasMapping;
    for (const Alias& alias : aliases) {
        aliasMapping[alias.Alias] = alias.PersonId;
    }

    std::unordered_map<int, std::string> personMapping;
    for (const Person& person : persons) {
        personMapping[person.Id] = person.Name;
    }

    // Create a vector of edges
    std::vector<Edge> edges;
    for (const Email& email : emails) {
        std::string sender = unifyName(email.MetadataFrom, aliasMapping, personMapping);
        std::string receiver = unifyName(email.MetadataTo, aliasMapping, personMapping);
        bool edgeFound = false;
        for (Edge& edge : edges) {
            if (edge.sender == sender && edge.receiver == receiver) {
                edge.weight++;
                edgeFound = true;
                break;
            }
        }
        if (!edgeFound) {
            edges.push_back({sender, receiver, 1});
        }
    }

    // Calculate PageRank
    std::map<std::string, double> pageRank = calculatePageRank(edges);

    // Output PageRank results
    for (const auto& pair : pageRank) {
        std::cout << "Node: " << pair.first << ", PageRank: " << pair.second << std::endl;
    }

    return 0;
}

File: file.py
import pandas as pd
import networkx as nx
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.graph_objects as go
import folium

# Load the emails dataset
emails = pd.read_csv("./input/Emails.csv")

# Load the aliases dataset
aliases = pd.read_csv("./input/Aliases.csv")
alias_mapping = dict(zip(aliases['Alias'], aliases['PersonId']))

# Load the persons dataset
persons = pd.read_csv("./input/Persons.csv")
person_mapping = dict(zip(persons['Id'], persons['Name']))

# Function to unify names and aliases
def unify_name(name):
    name = str(name).lower().split("@")[0].replace(",", "")
    return person_mapping.get(alias_mapping.get(name, name), name)

# Create a directed graph
graph = nx.DiGraph()

# Iterate through the emails dataset and add edges with weights
for _, row in emails.iterrows():
    sender = unify_name(row['MetadataFrom'])
    receiver = unify_name(row['MetadataTo'])
    if graph.has_edge(sender, receiver):
        graph[sender][receiver]['weight'] += 1
    else:
        graph.add_edge(sender, receiver, weight=1)

# Calculate PageRank for each node and set it as an attribute
pagerank = nx.pagerank(graph)
nx.set_node_attributes(graph, pagerank, 'pagerank')

# Function to display the network graph using Plotly
def show_graph(graph):
    edge_x = []
    edge_y = []
    for edge in graph.edges():
        x0, y0 = edge[0], edge[1]
        edge_x.append(x0)
        edge_y.append(y0)

    edge_trace = go.Scatter(
        x=edge_x, y=edge_y,
        line=dict(width=0.5, color='#888'),
        hoverinfo='none',
        mode='lines')

    node_x = []
    node_y = []
    node_text = []
    for node in graph.nodes():
        x, y = node, graph.nodes[node]['pagerank']
        node_x.append(x)
        node_y.append(y)
        node_text.append(f'{node}<br>PageRank: {y:.3f}')

    node_trace = go.Scatter(
        x=node_x, y=node_y,
        mode='markers',
        text=node_text,
        marker=dict(
            showscale=True,
            colorscale='YlGnBu',
            size=10,
            colorbar=dict(
                thickness=15,
                title='PageRank',
                xanchor='left',
                titleside='right'
            )
        )
    )

    fig = go.Figure(data=[edge_trace, node_trace],
                    layout=go.Layout(
                        showlegend=False,
                        hovermode='closest',
                        title="Hillary's Email Communication Network",
                        xaxis=dict(showgrid=False, zeroline=False),
                        yaxis=dict(showgrid=False, zeroline=False)
                    ))
    fig.show()

# Display the complete network graph using Plotly
show_graph(graph)

# Function to filter and display a simplified network graph using Plotly
def filter_and_show_graph(graph, threshold):
    filtered_graph = graph.copy()
    for node, data in graph.nodes(data=True):
        if data['pagerank'] < threshold:
            filtered_graph.remove_node(node)
    
    show_graph(filtered_graph)

# Display a simplified network graph with a pagerank threshold using Plotly
pagerank_threshold = 0.005
filter_and_show_graph(graph, pagerank_threshold)

# Data Visualization with Seaborn and Matplotlib
sns.set(style="whitegrid")
plt.figure(figsize=(10, 6))
sns.boxplot(x=[data['MetadataFrom'], data['MetadataTo']], y='pagerank', data=graph.nodes(data=True))
plt.xticks(rotation=45)
plt.title('Pagerank Distribution for Senders and Receivers')
plt.show()

# Create a Folium Map
m = folium.Map(location=[0, 0], zoom_start=1)
folium.Marker(location=[0, 0], tooltip="Hillary's Email Communication Network").add_to(m)
m.save("email_network_map.html")
